shelley-ma/impl/src/Cardano/Ledger/Mary/Value.hs

```diff
@@ -161,15 +167,55 @@ instance
 
 -- ========================================================================
 -- Compactible
--- This is used in the TxOut which stores the (CompactForm Value). For now that is just Value
--- We may revisit this, and make it some kind serialized bytes.
+-- This is used in the TxOut which stores the (CompactForm Value).
 
-instance Compactible (Value era) where
-  -- TODO a proper compact form
-  newtype CompactForm (Value era) = CompactValue {getCompactValue :: Value era}
+instance Era era => Compactible (Value era) where
+  newtype CompactForm (Value era) = CompactValue (CV era)
     deriving (ToCBOR, FromCBOR)
-  toCompact = CompactValue
-  fromCompact = getCompactValue
+  toCompact = CompactValue . toCV
+  fromCompact (CompactValue x) = fromCV x
+
+instance (Typeable (Core.Script era), Era era) => ToCBOR (CV era) where
+  toCBOR = toCBOR . fromCV
+
+instance (Typeable (Core.Script era), Era era) => FromCBOR (CV era) where
+  fromCBOR = toCV <$> fromCBOR
+
+data CV era
+  = CV
+      {-# UNPACK #-} !Word64
+      {-# UNPACK #-} !(Array Int (CVPart era))
+
+data CVPart era
+  = CVPart
+      (PolicyID era)
```

Comment:
Maybe unrelated to this PR, but does `PolicyID` really need to be parameterised over the `era`? The crypto should be enough, right?
```haskell
newtype PolicyID era = PolicyID {policyID :: ScriptHash era}

newtype ScriptHash era
  = ScriptHash (Hash.Hash (ADDRHASH (Crypto era)) (Core.Script era))
```

You could apply the `EraIndependentTxBody` trick here too?

Reply:
yes, crypto should be enough, I do think the `EraIndependentTxBody` trick would make this better. (I'd probably do this in a separate PR)